src/Compiler/Driver/ReuseTcResults/CachingDriver.fs

```diff
@@ -0,0 +1,139 @@
+
+    [<Literal>]
+    let CmdLineHeader = "CMDLINE"
+
+    [<Literal>]
+    let GraphHeader = "GRAPH"
+
+    [<Literal>]
+    let ReferencesHeader = "REFERENCES"
+
+    let writeThisTcData (tcData: TcData) =
+        use tcDataFile = FileSystem.OpenFileForWriteShim tcDataFilePath
+
+        let lines = ResizeArray<string>()
+        lines.Add $"BEGIN {CmdLineHeader}"
+        lines.AddRange tcData.CmdLine
+        lines.Add $"BEGIN {GraphHeader}"
+        lines.AddRange tcData.Graph
+        lines.Add $"BEGIN {ReferencesHeader}"
+        lines.AddRange tcData.References
+
+        tcDataFile.WriteAllLines lines
+
+    let readPrevTcData () =
+        if FileSystem.FileExistsShim tcDataFilePath then
+            use tcDataFile = FileSystem.OpenFileForReadShim tcDataFilePath
+
+            let cmdLine = ResizeArray<string>()
+            let graph = ResizeArray<string>()
+            let refs = ResizeArray<string>()
+
+            let mutable currentHeader = ""
+
+            tcDataFile.ReadLines()
+            |> Seq.iter (fun line ->
+                match line with
+                | line when line.StartsWith "BEGIN" -> currentHeader <- line.Split ' ' |> Array.last
+                | line ->
+                    match currentHeader with
+                    | CmdLineHeader -> cmdLine.Add line
+                    | GraphHeader -> graph.Add line
+                    | ReferencesHeader -> refs.Add line
+                    | _ -> invalidOp "broken tc cache")
+
+            Some
+                {
+                    CmdLine = cmdLine.ToArray()
+                    Graph = graph.ToArray()
+                    References = refs.ToArray()
+                }
+
+        else
+            None
+
+    let getThisCompilationCmdLine args = args
+
+    // maybe split into two things?
+    let getThisCompilationGraph inputs =
+        let sourceFiles =
+            inputs
+            |> Seq.toArray
+            |> Array.mapi (fun idx (input: ParsedInput) ->
+                {
+                    Idx = idx
+                    FileName = input.FileName
+                    ParsedInput = input
+                })
+
+        let filePairs = FilePairMap sourceFiles
+        let graph, _ = DependencyResolution.mkGraph filePairs sourceFiles
+
+        let list = List<string>()
+
+        for KeyValue(idx, _) in graph do
+            let fileName = sourceFiles[idx].FileName
+            let lastWriteTime = FileSystem.GetLastWriteTimeShim fileName
+            list.Add $"%i{idx},{fileName},{lastWriteTime.ToOADate()}"
+
+        for KeyValue(idx, deps) in graph do
+            for depIdx in deps do
+                list.Add $"%i{idx} --> %i{depIdx}"
+
+        list.ToArray()
+
+    let getThisCompilationReferences =
+        List.map (fun (r: AssemblyReference) -> r.Text)
+        >> List.map (fun fileName -> fileName, FileSystem.GetLastWriteTimeShim fileName)
+        >> List.map (fun (fileName, lastWriteTime) -> $"{fileName},{lastWriteTime.ToOADate()}")
```

Comment:
Why does this need to be an OLE date?

Reply:
Actually a few reasons:
- it's the easiest sufficient formatting out-of-the-box (it's there in `DateTime`)
- we need to be very precise here (seconds were not enough
- I like the fact it splits date and time parts

Reply:
Please, don't use OLE date, `System.DateTime.Now.Ticks` has needed precision, you shouldn't format it to string, due to difference in culture and how different tools set this culture up, it can happen that between two runs of compiler, it will produce different results of the same date.

Reply:
Alrighty, changed here and in the other instance.

Reply:
> System.DateTime.Now.Ticks has needed precision,

Just to add, other places in the compiler use the UTC time (at least, it was the case a few years ago, I haven't checked it since then, but I'd expect it to be maintained like this)

Reply:
Yeah, it's a bit confusing but our GetLastTimeWriteShim actually uses GetLastTimeWriteUtc under the cover :)

Reply:
> Yeah, it's a bit confusing but our GetLastTimeWriteShim actually uses GetLastTimeWriteUtc under the cover :)

Yeah, because there's no legit reason to use anything other than utc.